opal/mca/memory/patcher/memory_patcher_component.c

```diff
@@ -55,6 +55,18 @@
 #include "memory_patcher.h"
 #undef opal_memory_changed
 
+#if defined(SYS_shmdt) || (defined(IPCOP_shmdt) && defined(SYS_ipc))
```

Comment:
I just moved the old code up higher in the file so that `HAS_SHMAT` could be used earlier in the file.  I don't have much of an opinion on collapsing them into a single line or not.

Reply:
@bosilca and I talked this through on Slack.  It turns out that collapsing these `#define`s actually results in new warnings (probably sorta similar to the undefined behavior described in the commit message, but we didn't dig too deeply into it).  So we'll leave these `#define`s as they are for now.